webhook/resources/node/validator.go

```diff
@@ -0,0 +1,149 @@
+package node
+
+import (
+	"fmt"
+	"math"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+	"github.com/sirupsen/logrus"
+)
+
+type nodeValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &nodeValidator{ds: ds}
+}
+
+func (n *nodeValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "nodes",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Node{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Update,
+			admissionregv1.Delete,
+		},
+	}
+}
+
+func (n *nodeValidator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) error {
+	oldNode := oldObj.(*longhorn.Node)
+	newNode := newObj.(*longhorn.Node)
+
+	// Only scheduling disabled node can be evicted
+	// Can not enable scheduling on an evicting node
+	if oldNode.Spec.EvictionRequested && oldNode.Spec.AllowScheduling {
+		return werror.NewInvalidError(fmt.Sprintf("need to disable scheduling on node %v for node eviction, or cancel eviction to enable scheduling on this node",
+			oldNode.Name), "")
+	}
+
+	// We need to make sure the tags passed in are valid before updating the node.
+	_, err := util.ValidateTags(newNode.Spec.Tags)
+	if err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if newNode.Spec.EngineManagerCPURequest < 0 || newNode.Spec.ReplicaManagerCPURequest < 0 {
```

Comment:
Don't these have been covered by OpenAPI schema validation? `+kubebuilder:validation:Minimum`

Reply:
Yes, we can remove the checks.